### PR TITLE
Feature/database

### LINKED ISSUE #1: Database Design and Creation
## Description

The goal for this phase of development is to design and develop a sample database for eva-gram. This sample database will be used for later backend and frontend development phases and will follow a schema that will be used in the database hosted on discover. Here is the process for completing this phase:

- [x] Download and review sample experiment diagnostics from Box
- [x] Create a postgres server and sample database on personal computer
- [x] Develop python script to loop through all pkl files, extract the div and script components from pickled dictionaries, and save it to the sample database. Information to include in the database entry: username, experiment id, observation type (names of the directories such as amsua_n19, aircraft, etc.)
- [x] Create database schema documentation for sample database (UML diagram)
- [x] Develop test script for the github action that will call the database input tool with a given sample dataset. Wire it to local database and see that it works as expected.
- [x] Devise a github action to test tool to make sure the database input mechanism works as expected
- [x] Create and merge PR for tool, github action, and sample data to use during the github action

## Acceptance Criteria (Definition of Done)

A tool that takes in the location of eva hvplot figures, converts the figures to it's div and script components, and saves components to a postgres database. A github action that creates a database and runs the tool to check if it works as expected is complete.

### PR DESCRIPTION
**Pull request by brandonzhu09**
## Description

Developed a database input tool that creates a connection with the PostgreSQL database and stores dynamic Eva hvplot components under a database schema from data input sources like pickle files and JSON datasets.  

## Issue(s) addressed

Resolves #1

### CODE REVIEW

File: __pycache__/input_tool.cpython-311.pyc
Comment (ghost) [inline __pycache__/input_tool.cpython-311.pyc:1]: add pycache to .gitignore

File: input_tool.py
Comment (ghost) [inline input_tool.py]: Is there a schema file in this repo?  I know we'll be using django models later, but we should have the schema stored and maintained in a production .sql file first. 

Also, are you using a postgres desktop client to configure the db, or are you doing it all by hand?

File: input_tool.py
Comment (ghost) [inline input_tool.py]: Should `experiment_name` be not null and unique?  And any other `_name` fields for that matter?

**Comment by ghost** [comment]
@asewnath Can you please create branch protection rules to require approval before merging.

EDIT:  Thank you

File: __pycache__/input_tool.cpython-311.pyc
Comment (asewnath) [reply __pycache__/input_tool.cpython-311.pyc:1]: Adding .gitignore file

Comment (asewnath) [reply __pycache__/input_tool.cpython-311.pyc:1]: @brandonzhu09 please remove pycache directory

File: local/script.txt
Comment (asewnath) [inline local/script.txt:1]: remove

File: local/script-tag.txt
Comment (asewnath) [inline local/script-tag.txt:1]: remove

File: local/script-demo.txt
Comment (asewnath) [inline local/script-demo.txt:1]: remove

File: local/plot_ids_file.txt
Comment (asewnath) [inline local/plot_ids_file.txt:1]: remove

File: local/plot_1.pkl
Comment (asewnath) [inline local/plot_1.pkl:1]: remove

File: local/dataset_ids.txt
Comment (asewnath) [inline local/dataset_ids.txt:1]: Is this file needed? If so, it needs to be out of this directory

File: __pycache__/input_tool.cpython-311.pyc
Comment (brandonzhu09) [reply __pycache__/input_tool.cpython-311.pyc:1]: Ok, will do that.

File: local/airTemperature.pkl
Comment (asewnath) [inline local/airTemperature.pkl:1]: Does this need to be in the directory?

File: local/amsua_metop-b/brightnessTemperature13_a_hofx_vs_gsihofx.pkl
Comment (asewnath) [inline local/amsua_metop-b/brightnessTemperature13_a_hofx_vs_gsihofx.pkl:1]: Is metop-b used for testing? If not we can remove it

File: local/amsua_aqua/brightnessTemperature_10_b_hofx-vs-gsihofxbc.pkl
Comment (asewnath) [inline local/amsua_aqua/brightnessTemperature_10_b_hofx-vs-gsihofxbc.pkl:1]: `local/` should be changed to `eva/` to mimic how the dataset would appear to the input_tool. The only thing that should be in `eva/` are the observation directories containing the pickle files. You can also remove the `b` from the pkl file name (you can remove all single letters like `b` from the other pkl files).

File: input_tool.py
Comment (asewnath) [inline input_tool.py]: change `'local'` to `'eva'` to reflect change to data directory

Comment (asewnath) [reply input_tool.py]: I would also think about passing the path to `eva/` into the input tool since it won't always be in the current working directory

File: input_tool.py
Comment (asewnath) [inline input_tool.py]: How will `dataset.json` be automatically generated? When the input tool is used in the context of swell, it will be given a user, experiment_name, and the location of the eva directory with pkl files

File: input_tool.py
Comment (asewnath) [inline input_tool.py]: Do we need a `UNIQUE` in this table?

**Review by asewnath — CHANGES_REQUESTED** [review]

File: input_tool.py
Comment (ghost) [reply input_tool.py]: Like a unique on `(experiment_id, group_id, and observation_id)` ?

**Comment by asewnath** [comment]
Great work @brandonzhu09! This is a huge step in eva-gram's development. In an effort to keep organized for future features, let's clean up the file structure. Maybe we can try something like this?

Screenshot 2024-01-11 at 2 04 52 PM

Down the line we'd probably want to isolate all the tests in it's own directory and put all the other code in `src/eva-gram/src`, but I don't think we need to be concerned with that just yet.

File: input_tool.py
Comment (brandonzhu09) [reply input_tool.py]: I have the experiment path point to the `local` directory temporarily until all the pkl filenames in `eva` have changed to reflect the new naming convention. Do you know where I can download the new directory with the filename changes?

Comment (asewnath) [reply input_tool.py]: You can still use the experiment path pointing to the `local` directory, just change `local` to `eva` since the files in your `local` directory should be updated with the new naming convention. You would be passing the contents of `EXPERIMENT_DATA_PATH` into the tool

Comment (brandonzhu09) [reply input_tool.py]: Got it, thanks!

**Comment by ghost** [comment]
I'd make a `dir` for `.sql` files.  We could end up accruing many stored procedures over time.

File: local/amsua_aqua/brightnessTemperature_10_b_hofx-vs-gsihofxbc.pkl
Comment (brandonzhu09) [reply local/amsua_aqua/brightnessTemperature_10_b_hofx-vs-gsihofxbc.pkl:1]: Because of the way my program parses the filenames, removing `b` will create an out-of-bounds error since I treated it as an additional parameter. This is a simple fix on my end but are we sure that this single character in the filename won't appear in production pkl files?

File: input_tool.py
Comment (brandonzhu09) [reply input_tool.py]: Yes, I can reflect this in the schema.

File: input_tool.py
Comment (brandonzhu09) [reply input_tool.py]: For experiment_name, I added the unique constraint `(experiment_name, owner_id)` as my reasoning was that experiment name should be unique for the same owner but can be the same for multiple owners.

Comment (brandonzhu09) [reply input_tool.py]: I will add the not null constraint for `experiment_name` though.

Comment (ghost) [reply input_tool.py]: Yeah I think that's right.

File: local/amsua_aqua/brightnessTemperature_10_b_hofx-vs-gsihofxbc.pkl
Comment (asewnath) [reply local/amsua_aqua/brightnessTemperature_10_b_hofx-vs-gsihofxbc.pkl:1]: Yeah we can remove the single character upstream in eva

File: input_tool.py
Comment (ghost) [inline input_tool.py]: Why are these commands identical?

Comment (ghost) [reply input_tool.py]: In today's meeting we agreed to ignore first/last name for now (allow NULL), but to enforce not-null values for experiment, group, variable, and observation names.

File: input_tool.py
Comment (ghost) [inline input_tool.py]: Please put a comment here so the reader immediately knows what attribute each plot_component corresponds to

**Comment by ghost** [comment]
In issue #1 we settled on a class diagram for owner, experiment, etc..  This PR expresses it in SQL, but no Python classes are defined for it.  I was expecting to have an object-oriented approach to the API, where each SQL table/record is represented by a Python class/object.  Is that indeed a plan or not?

**Comment by asewnath** [comment]
> In issue #1 we settled on a class diagram for owner, experiment, etc.. This PR expresses it in SQL, but no Python classes are defined for it. I was expecting to have an object-oriented approach to the API, where each SQL table/record is represented by a Python class/object. Is that indeed a plan or not?

@tariqhamzeyssai we should replace that class diagram with the schema Brandon generates from pgadmin

**Comment by brandonzhu09** [comment]
> In issue #1 we settled on a class diagram for owner, experiment, etc..  This PR expresses it in SQL, but no Python classes are defined for it.  I was expecting to have an object-oriented approach to the API, where each SQL table/record is represented by a Python class/object.  Is that indeed a plan or not?  

Yes, that is indeed the plan. I have began work on the API in the Django backend directory and there is a separate `models.py` file that stores all the existing tables from the database in Model objects.

**Comment by ghost** [comment]
> Model objects

Right.  I look forward to seeing that.  As such, I'll withhold some comments until we dive into django world.

**Comment by ghost** [comment]
One more comment:  It would be nice if queries were constructed in a more concise way.  Here's some pseudo code that does it dynamically, and will handle conditions where null or missing values are acceptable.

```
query_string = INSERT INTO {table} 
dictionary.keys() 
VALUES
%s for length(dictionary.keys()
```

This assumes that the dictionary only contains information relevant to the query.  You can create a method for generating these query strings and then call it in every `add` function.  A lot of space will be saved and the code will be easier to read.

**Comment by asewnath** [comment]
@brandonzhu09 If you take a look at the develop branch, you'll see src and tests directories. We can organize your code in these directories like so:

Screenshot 2024-01-16 at 10 09 35 AM

**Comment by brandonzhu09** [comment]
> @brandonzhu09 If you take a look at the develop branch, you'll see src and tests directories. We can organize your code in these directories like so:
> 
> Screenshot 2024-01-16 at 10 09 35 AM

Will reflect the new directory structure in the next commit. Thanks for the outline!

**Comment by asewnath** [comment]
@brandonzhu09 for the eva directory, we don't need all the observation files in this repository, just the few that you were testing with

File: src/eva-gram/src/database/input_tool.py
Comment (ghost) [inline src/eva-gram/src/database/input_tool.py]: This might not be relevant once we move to django, but a cleaner way of doing this is could be like:

```
required = ['plot_id', 'plot_file', 'experiment_id', 'group_id', 'observation_id']
difference = set(required).difference(plot_obj)
if len(difference) > 0:
    print(f'Missing required keys for plot: {difference}')
```
and perhaps it could be generalized in `insert_table_record`.

File: src/eva-gram/src/database/input_tool.py
Comment (ghost) [inline src/eva-gram/src/database/input_tool.py]: Can we not use `insert_table_record` here?

File: src/eva-gram/src/database/input_tool.py
Comment (ghost) [inline src/eva-gram/src/database/input_tool.py]: Might have been discussed before, but can file names reflect null channel instead of hardcoding the case into Python?  For example:

`specificHumidity__hofx-vs-gsihofx.pkl`
and note the double delimiter.

**Comment by ghost** [comment]
Thanks @brandonzhu09 for addressing our comments!  This PR is looking much better now.  We'll talk later and, unless @asewnath has additional concerns, I think this PR suffices for the inaugural merge.

**Comment by brandonzhu09** [comment]
> @brandonzhu09 for the eva directory, we don't need all the observation files in this repository, just the few that you were testing with

@asewnath my bad, making the change in the next commit

**Comment by climbfuji** [comment]
@brandonzhu09 Please note :

> GITHUB PUSH ERROR: The size of this repository, 370.996MB, exceeds the limit of 50.0MB. 
> Please arrange for large files to be stored on git-LFS.
>
> Repo: eva-gram
> Branch: feature/database
> User: brandonzhu09
> User email: ...

Can you please do that (may need to rewrite the history of your PR to get rid of the large files that are not on git-LFS)?Thanks!

**Comment by asewnath** [comment]
@brandonzhu09 Hey brandon, be careful with "git pull" right now. I had to rewrite the commit history in order to remove the eva pkl files, and unfortunately it was messy and a lot of the commits you had made probably aren't here anymore. When you have the time, could you go back and add the changes you expect to see in this branch? I would make a new clone on your computer, checkout your branch, add your changes from the clone you're working in, and then push the changes to fix the branch. Sorry about this!

**Comment by asewnath** [comment]
Also while I was doing this I removed the extra src directory after src/evagram/. We typically don't need the extra src directory, that was a mistake on my part

**Comment by brandonzhu09** [comment]
> Also while I was doing this I removed the extra src directory after src/evagram/. We typically don't need the extra src directory, that was a mistake on my part

@asewnath Thanks for the heads up! Sorry about the commit history, I know it has to do with the observation files I pushed to the repository. I will make a commit soon to restore the changes I made earlier, meanwhile, is it safe to delete the extra src folder in src/evagram/?

**Comment by asewnath** [comment]
> > Also while I was doing this I removed the extra src directory after src/evagram/. We typically don't need the extra src directory, that was a mistake on my part
> 
> @asewnath Thanks for the heads up! Sorry about the commit history, I know it has to do with the observation files I pushed to the repository. I will make a commit soon to restore the changes I made earlier, meanwhile, is it safe to delete the extra src folder in src/evagram/?

I think so, I moved all the stuff from src/eva-gram/src up one level. The only thing that was in there was the database folder

**Comment by asewnath** [comment]
I have fixes for properly importing the module so that we can get this PR to pass the coding norms test. I can add those to this branch once the PR is back to the state it was at before the commit history clean up

**Comment by brandonzhu09** [comment]
> I have fixes for properly importing the module so that we can get this PR to pass the coding norms test. I can add those to this branch once the PR is back to the state it was at before the commit history clean up

@asewnath I will make that commit soon, there is a big refactor I wanted to include but once that is done you can make the fix.

**Review by asewnath — APPROVED** [review]
Good job @brandonzhu09! First PR is ready to be merged! Make sure to choose squash and merge

**Comment by brandonzhu09** [comment]
> Good job @brandonzhu09! First PR is ready to be merged! Make sure to chose squash and merge

Ok, thank you!

**Pull request merged**
